File: 13/13.pl
#!/usr/bin/perl

#
# Copyright (C) 2023 Joelle Maslak
# All Rights Reserved - See License
#

use JTM::Boilerplate 'script';

use lib '.';
use Table;

use List::Util qw(min);

MAIN: {
    my $t = Table->new();
    my $rowcount = 0;
    my $sum_a = 0;
    my $sum_b = 0;
    while (my $line = <<>>) {
        chomp($line);
        if ($line eq "") {
            $sum_a = $sum_a + get_smudged($t);
            $sum_b = $sum_b + get_unsmudged($t);
            $t = Table->new();
            $rowcount = 0;
        } else {
            my $colcount = 0;
            for my $c (split //, $line) {
                $t->put_xy($rowcount, $colcount, $c);
                $colcount++;
            }
            $rowcount++;
        }
    }
    $sum_a = $sum_a + get_smudged($t);
    $sum_b = $sum_b + get_unsmudged($t);

    say "Sum part A: $sum_a";
    say "Sum part B: $sum_b";
}

sub get_smudged($t1) {
    my $t2 = $t1->copy_swap_xy();
    my $sum = get_values($t1)->[0] + get_values($t2)->[0] * 100;
    return $sum;
}

sub get_unsmudged($t) {
    my $rows = $t->row_count();
    my $cols = $t->col_count();

    my $t1 = $t->copy();
    my $t2 = $t->copy_swap_xy();

    my $orig1 = get_values($t1);
    my $orig2 = get_values($t2);

    for (my $i=0; $i<$rows; $i++) {
        for (my $j=0; $j<$cols; $j++) {
            $t1->put_xy($i, $j, $t1->get_xy($i, $j) eq '.' ? '#' : '.' );
            $t2->put_xy($j, $i, $t2->get_xy($j, $i) eq '.' ? '#' : '.' );

            my $r1 = get_values($t1, [$orig1->[1], $orig1->[2]])->[0];
            my $r2 = get_values($t2, [$orig2->[1], $orig2->[2]])->[0];

            if ($r1 and $r2) {
                return $r1 + $r2 * 100;
            }
            if ($r1 and $r1 != $orig1->[0]) {
                return $r1;
            } elsif ($r2 and $r2 != $orig2->[0]) {
                return $r2 * 100;
            }

            $t1->put_xy($i, $j, $t1->get_xy($i, $j) eq '.' ? '#' : '.' );
            $t2->put_xy($j, $i, $t2->get_xy($j, $i) eq '.' ? '#' : '.' );
        }
    }

    return 0
}

sub get_values($t, $ignore = [-1, -1]) {
    my $cols = $t->col_count();
    for (my $i=1; $i<$cols; $i++) {
        my $dist = min ( $i, $cols-$i );
        my $left = $i-$dist;
        my $right = $i+$dist-1;

        if ($left == $ignore->[0] and $right == $ignore->[1]) { next; }

        my $flag=1;
        for (my $x=1; $x<=$dist; $x++) {
            my ($l) = join "", $t->get_col($left+$x-1);
            my ($r) = join "", $t->get_col($right-$x+1);

            if ($l ne $r) { $flag = undef; }
        }

        if ($flag) {
            return [$i, $left, $right];
        }
    }
    return [0, -1, -1];
}

File: 13/13.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	var t Table[rune]

	sumA := 0
	sumB := 0
	scanner := bufio.NewScanner(os.Stdin)
	rowNumber := 0
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) == 0 {
			sumA += getSmudged(t)
			sumB += getUnsmudged(t)
			t.Clear()
			rowNumber = 0
		} else {
			cols := []rune(line)
			for colNumber, v := range cols {
				t.PutXY(rowNumber, colNumber, v)
				colNumber++
			}
			rowNumber++
		}
	}
	sumA += getSmudged(t)
	sumB += getUnsmudged(t)

	fmt.Printf("Sum part A: %d\n", sumA)
	fmt.Printf("Sum part B: %d\n", sumB)
}

func getSmudged(t1 Table[rune]) int {
	t2 := t1.CopySwapXY()
	val1, _, _ := getValues(t1, -1, -1)
	val2, _, _ := getValues(t2, -1, -1)
	return val1 + val2*100
}

func getUnsmudged(t Table[rune]) int {
	rows := t.RowCount()
	cols := t.ColCount()

	t1 := t.Copy()
	t2 := t.CopySwapXY()

	_, small1, big1 := getValues(t1, -1, -1)
	_, small2, big2 := getValues(t2, -1, -1)

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			c := t1.GetXY(i, j)
			var newC rune
			if c == '.' {
				newC = '#'
			} else {
				newC = '.'
			}

			t1.PutXY(i, j, newC)
			t2.PutXY(j, i, newC)

			r1, _, _ := getValues(t1, small1, big1)
			r2, _, _ := getValues(t2, small2, big2)

			if r1 > 0 || r2 > 0 {
				return r1 + r2*100
			}

			t1.PutXY(i, j, c)
			t2.PutXY(j, i, c)
		}
	}
	return 0
}

func getValues(t Table[rune], small int, big int) (int, int, int) {
	cols := t.ColCount()
	for i := 1; i < cols; i++ {
		var dist int
		if i < cols-i {
			dist = i
		} else {
			dist = cols - i
		}
		left := i - dist
		right := i + dist - 1

		if left == small && right == big {
			continue
		}

		flag := true
		for x := 1; x <= dist; x++ {
			l := t.GetCol(left + x - 1).GetRows()
			r := t.GetCol(right - x + 1).GetRows()

			if string(l) != string(r) {
				flag = false
			}
		}
		if flag {
			return i, left, right
		}
	}
	return 0, -1, -1
}
